### ISSUE TITLE
Any suggestions for adding Punctuations and fullstops?

### ISSUE THREAD

**Issue by Caet-pip** [k2-fsa/sherpa-onnx#291]
Hello I am currently using the sherpa-onnx-streaming-zipformer-en-2023-06-26 and sherpa-onnx-streaming-zipformer-en-2023-06-21 models. The output is great but lacks punctuation. Is there any way to add these, any suggestions would be appreciated thanks.

**Comment by csukuangfj** [comment]
You may find the following link helpful.

https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/nlp/punctuation_and_capitalization.html

**Comment by danpovey** [comment]
I think it would be easier and much more accurate to simply use a version of the Zipformer that has been trained with punctuation.  Fangjun, have we not uploaded such a model yet?

**Comment by csukuangfj** [comment]
I think @marcoyang1998  xiaoyu has not uploaded such a model, right?

**Comment by Caet-pip** [comment]
Is there a punctuation model which can be used in sherpa-onnx?

**Comment by Caet-pip** [comment]
> https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/nlp/punctuation_and_capitalization.html

Can this be used with Sherpa-onnx Zipformer models

**Comment by csukuangfj** [comment]
> Is there a punctuation model which can be used in sherpa-onnx?

Yes, models from https://github.com/k2-fsa/icefall/pull/1099
can be used in sherpa-onnx.

@marcoyang1998  Do you think we can first release a model with punctuation and capitalization support?

**Comment by csukuangfj** [comment]
> > https://docs.nvidia.com/deeplearning/nemo/user-guide/docs/en/main/nlp/punctuation_and_capitalization.html
> 
> Can this be used with Sherpa-onnx Zipformer models

I am afraid they are not supported in sherpa-onnx.

**Comment by Caet-pip** [comment]
> Yes, models from [k2-fsa/icefall#1099](https://github.com/k2-fsa/icefall/pull/1099)
> can be used in sherpa-onnx.

Is there any way to export this model right now? I cant find the model in HuggingFace

**Comment by pkufool** [comment]
> > Is there a punctuation model which can be used in sherpa-onnx?
> 
> Yes, models from [k2-fsa/icefall#1099](https://github.com/k2-fsa/icefall/pull/1099) can be used in sherpa-onnx.
> 
> @marcoyang1998 Do you think we can first release a model with punctuation and capitalization support?

I have a model trained on libriheavy large, will upload to huggingface.

**Comment by pkufool** [comment]
@Caet-pip I have uploaded my latest model to https://huggingface.co/pkufool/icefall-asr-zipformer-libriheavy-punc-20230830 , the training data is here https://huggingface.co/datasets/pkufool/libriheavy .  Sorry, tha whole work has not been  finished yet, so some of them might be lack of documents, if you want to know how we construct the libriheavy dataset, you can see this https://github.com/k2-fsa/text_search/tree/master/examples/libriheavy first. 
Actually, the training is still running, the model I uploaded is converted from epoch 10 checkpoint, will update to a better one (more epochs) later.

Here is the decoding results:
for the first row, the transcripts are with cases and punctuations, also the decoded text.
for the second row, I normalized both of the transcripts and decoded text (removing the punctuations and converting to uppper case). 

Decoding text type | Librispeech WERs(clean & other) | Libriheavy WERs(clean & other)
-- | -- | --
Cases with punc | - | 8.21 & 11.76
Upper no punc | 1.77 & 3.76 | 2.37 & 5.68

**Comment by csukuangfj** [comment]
Could you add some test waves and show some example out with it using sherpa-onnx?

**Comment by pkufool** [comment]
The libriheavy audios are long, will finally upload some test waves, but not now, I think @Caet-pip  know how to use our model in sherpa-onnx.

Here are some decoding results from icefall:
```
My heart runs o'er with sweet surmises, The while my fancy weaves her rhyme; Kind hopes and musical surprises Throng round me from the olden time. 

Alas! too soon the veil is lifted That hangs between the soul and pain; Too soon the morning red hath drifted Into dull cloud or fallen in rain.

I needs must droop my head And brush away a happy tear, For they are gone, and dry and dead The rosebud lies before me here. Yet  it is in no stranger's hand, For I will guard it tenderly, And it shall be a magic wand To bring mine own true love to me.

A white headed negro man with spectacles and damaged white cotton gloves appeared in the presence, made a stately obeisance, and announced: "Marse Washington Hawkins, suh."
```

**Comment by Caet-pip** [comment]
Thank you so much! @pkufool, @csukuangfj and @danpovey 

I will try converting this model to sherpa-onnx using this guide in documentation. 
https://k2-fsa.github.io/icefall/model-export/export-onnx.html#export-the-model-to-onnx

**Comment by Caet-pip** [comment]
Thats great! by the way I wanted to ask this earlier but forgot, does this model support streaming transcription. If not do you have any models with streaming transcription and punctuation?

**Comment by csukuangfj** [comment]
please try it with a  non-streaming server.

**Comment by csukuangfj** [comment]
In general, if you don't find `streaming` in the repo name, it is a non-streaming model.

**Comment by Caet-pip** [comment]
Okay thank you very much, is there any streaming model which has punctuation support?

**Comment by pkufool** [comment]
> Okay thank you very much, is there any streaming model which has punctuation support?

No, we only trained a non-streaming model, you can either train by you self (I posted the dataset above) or wait for ours (around 1 month,  our machines are busy now).

**Comment by ezerhouni** [comment]
@pkufool 
When checking the tokens.txt I am seeing these tokens:
```
<0x00> 3
<0x01> 4
<0x02> 5
<0x03> 6
<0x04> 7
<0x05> 8
<0x06> 
...
```
Where do they come from ?

**Comment by marcoyang1998** [comment]
@ezerhouni We are using byte-level fallback when building the BPE.

**Comment by daiyangyang945** [comment]
@csukuangfj Hello, is there a Chinese punctuation model？

**Comment by pkufool** [comment]
> @csukuangfj Hello, is there a Chinese punctuation model？

Do you mean models that can output text and punctuations? If so, currently no.

**Comment by csukuangfj** [comment]
> Is there a punctuation model which can be used in sherpa-onnx?

@Caet-pip 

Yes, we just added one.

Please see the doc at
https://k2-fsa.github.io/sherpa/onnx/punctuation/index.html

The model supports only Chinese and English, however.
